### PR TITLE
4441 rel 6 4 bad references creation bug

### LINKED ISSUE #4441: bug - client is able to create resources with bad references despite configuration
**Describe the bug**

Client is able to create resources with bad references despite configuration.

**To Reproduce**

1. Enable the following DaoConfig properties:
    1. EnforceReferentialIntegrityOnWrite
    2. EnforceReferenceTargetTypes
    3. AutoCreatePlaceholderReferenceTargets
2. POST the example Patient below:

```json
{
  "resourceType": "Patient",
  "managingOrganization": {
    "reference": "urn:uuid:d8080e87-1842-46b4-aea0-b65803bc2897"
  }
}
```

Note the invalid [reference](https://hl7.org/fhir/references-definitions.html#Reference.reference).

**Observed behavior**

Server creates the resource. It will look something like:

```json
{
  "resourceType": "Patient",
  "id": "402",
  "meta": {
    "versionId": "1",
    "lastUpdated": "2023-01-17T21:33:53.525+00:00",
    "source": "#3HqrFGhiqa0m9oXs"
  },
  "managingOrganization": {
    "reference": "urn:uuid:d8080e87-1842-46b4-aea0-b65803bc2897"
  }
}
```

No reference target is created.

**Expected behavior**

- Server should not create the resource nor should it create a reference target.
- Server should issue an appropriate error.

**Screenshots**

**Environment (please complete the following information):**

- HAPI FHIR // 6.2.1 / build 71d8f80abf 2022-11-16

### PR DESCRIPTION
**Pull request by TipzCM**
Closes #4441

### CODE REVIEW

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/index/SearchParamWithInlineReferencesExtractor.java [hunk 1]
```diff
@@ -298,4 +336,53 @@ public void storeUniqueComboParameters(ResourceIndexedSearchParams theParams, Re
 			}
 		}
 	}
+
+	private class ExtractInlineReferencesParams {
```
Comment (StevenXLi) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/index/SearchParamWithInlineReferencesExtractor.java:340]: Question, is this the convention? should we keep method params classes within the same file?

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/index/SearchParamWithInlineReferencesExtractor.java [hunk 2]
```diff
@@ -188,16 +190,32 @@ public void setContext(FhirContext theContext) {
...
 	/**
-	 * Handle references within the resource that are match URLs, for example references like "Patient?identifier=foo". These match URLs are resolved and replaced with the ID of the
+	 * Handle references within the resource that are match URLs, for example references like "Patient?identifier=foo".
+	 * These match URLs are resolved and replaced with the ID of the
 	 * matching resource.
+	 *
+	 * This method is *only* called from UPDATE path
 	 */
-	public void extractInlineReferences(IBaseResource theResource, TransactionDetails theTransactionDetails, RequestDetails theRequest) {
+	public void extractInlineReferences(ExtractInlineReferencesParams theParams) {
```
Comment (jmarchionatto) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/index/SearchParamWithInlineReferencesExtractor.java:209]: todo: this method was already too long before, and worst now. Please split for readability

File: hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/provider/r4/ResourceProviderR4Test.java [hunk 3]
```diff
@@ -7556,6 +7555,113 @@ private void verifySinceBehaviourWhenQueriedDateBeforeTwoUpdatedDates(Long patie
...
+			// ensures target references are using the correct resource type
+			myDaoConfig.setEnforceReferenceTargetTypes(theInput.IsEnforceRefOnType);
+			// will create the resource if it does not already exist
+			myDaoConfig.setAutoCreatePlaceholderReferenceTargets(theInput.IsAutoCreatePlaceholderReferences);
+
+			// should fail
+			DaoMethodOutcome result = myPatientDao.create(patient, new SystemRequestDetails());
+
+			// a bad reference can never create a new resource
+			{
```
Comment (jmarchionatto) [inline hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/provider/r4/ResourceProviderR4Test.java:7640]: question: all this code before the catch is not expected to be reached. Or is it?

**Review by jmarchionatto — CHANGES_REQUESTED** [review]
Added a couple of comments

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/index/SearchParamWithInlineReferencesExtractor.java (hunk 1 above)
Comment (TipzCM) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/index/SearchParamWithInlineReferencesExtractor.java:319]: I don't like to, no.

I'll make a separate object

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/index/SearchParamWithInlineReferencesExtractor.java (hunk 2 above)
Comment (TipzCM) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/index/SearchParamWithInlineReferencesExtractor.java:209]: Removed code from here anyways

**Review by jmarchionatto — APPROVED** [review]

**Pull request merged**